core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,351 @@
+                        threadName(record.getThreadID()),
+                        record.getMessage());
+                  }
+
+                  private String threadName(long threadID) {
+                    Set<Thread> threadSet = Thread.getAllStackTraces().keySet();
+                    for (Thread t : threadSet) {
+                      if (t.getId() == threadID) {
+                        return t.getName();
+                      }
+                    }
+                    return "unknown";
+                  }
+                });
+            setLevel(Level.ALL);
+          }
+        };
+
+    Logger.getLogger("").addHandler(handler);
+    Logger l = Logger.getLogger(CloudSqlInstance.class.getName());
+    l.setLevel(Level.ALL);
+    logger.info("Hello");
+  }
+
+  private static class TestDataSupplier implements InstanceDataSupplier {
+
+    private final boolean flakey;
+
+    private AtomicInteger counter = new AtomicInteger();
+    private InstanceData response =
+        new InstanceData(
+            new Metadata(
+                ImmutableMap.of(
+                    "PUBLIC", "10.1.2.3",
+                    "PRIVATE", "10.10.10.10",
+                    "PSC", "abcde.12345.us-central1.sql.goog"),
+                null),
+            new SslData(null, null, null),
+            Date.from(Instant.now().plus(1, ChronoUnit.HOURS)));
+    private final ListeningScheduledExecutorService executor;
+
+    private TestDataSupplier(boolean flakey, ListeningScheduledExecutorService executor) {
+      this.flakey = flakey;
+      this.executor = executor;
+    }
+
+    @Override
+    public InstanceData getInstanceData(
+        CloudSqlInstanceName instanceName,
+        AccessTokenSupplier accessTokenSupplier,
+        AuthType authType,
+        ListeningScheduledExecutorService executor,
+        ListenableFuture<KeyPair> keyPair)
+        throws ExecutionException, InterruptedException {
+
+      ListenableFuture<InstanceData> f =
+          executor.submit(
+              () -> {
+                int c = counter.incrementAndGet();
+                Thread.sleep(100);
+
+                if (flakey && c % 2 == 0 && c > 10) {
+                  Thread.sleep(500);
+                  throw new ExecutionException("Flaky", new Exception());
+                }
+
+                return response;
+              });
+
+      return f.get();
+    }
+  }
+
+  private static class TestCredentialFactory implements CredentialFactory, HttpRequestInitializer {
+
+    @Override
+    public HttpRequestInitializer create() {
+      return this;
+    }
+
+    public void initialize(HttpRequest var1) throws IOException {
+      // do nothing
+    }
+  }
+
+  @Test
+  public void testCloudSqlInstanceConcurrency() throws Exception {
+    // Run the test 10 times to ensure we don't have race conditions
+    for (int i = 0; i < 10; i++) {
+      runConcurrencyTest();
+    }
+  }
+
+  @Test
+  public void testBasicHappyPath() throws Exception {
```

Comment:
Rather than happy path, let's make it descriptive: "testCloudSqlInstanceDoesXYZ"